src/OpenTelemetry/Metrics/MeterSdk.cs

```diff
@@ -52,17 +52,40 @@ public void Collect()
             {
                 // collect all pending metric updates and send to batcher.
                 // must sync to prevent multiple Collect occuring at same time.
+                var boundInstrumentsToRemove = new List<LabelSet>();
                 foreach (var longCounter in this.longCounters)
                 {
                     var metricName = longCounter.Key;
                     var counterInstrument = longCounter.Value;
                     foreach (var handle in counterInstrument.GetAllBoundInstruments())
                     {
                         var labelSet = handle.Key;
-                        var aggregator = handle.Value.GetAggregator();
+                        var aggregator = handle.Value.GetAggregator();                        
                         aggregator.Checkpoint();
                         this.metricProcessor.Process(this.meterName, metricName, labelSet, aggregator);
+
+                        if (handle.Value.Status == RecordStatus.CandidateForRemoval)
+                        {
+                            // It is possible that record status got promoted from a parallel thread.
+                            // The actual removal doesn't occur here.                            
+                            boundInstrumentsToRemove.Add(labelSet);
```

Comment:
I don't see the need to place the handle LabelSet's into the temporary list `boundInstrumentsToRemove`. There is no coordination between entries, so could we just call `Unbind` directly here?

Reply:
Calling `Unbind` would cause an entry to be deleted from the dictionary, while were are iterating over the dictionary. Thats why I collected the list, and called UnBind outside the iteration.